discovery-provider/src/queries/get_tips.py

```diff
@@ -0,0 +1,288 @@
+import logging
+from datetime import datetime
+from typing import List, Tuple, TypedDict, Union, cast
+
+from sqlalchemy import func, or_
+from sqlalchemy.orm import Query, aliased
+from sqlalchemy.orm.session import Session
+from src.models import AggregateUser, AggregateUserTips, Follow, User, UserTip
+from src.queries.get_unpopulated_users import get_unpopulated_users
+from src.queries.query_helpers import paginate_query, populate_user_metadata
+from src.utils.db_session import get_db_read_replica
+
+logger = logging.getLogger(__name__)
+
+
+class GetTipsArgs(TypedDict):
+    limit: int
+    offset: int
+    user_id: int
+    receiver_min_followers: int
+    receiver_is_verified: bool
+    current_user_follows: str
+    unique_by: str
+    min_slot: int
+    tx_signatures: List[int]
+
+
+class TipResult(TypedDict):
+    amount: int
+    sender: str
+    receiver: str
+    slot: int
+    created_at: datetime
+    followee_supporters: List[str]
+    tx_signature: str
+
+
+# Example of query with inputs:
+# limit=100, offset=0, user_id=1, current_user_follows=sender_or_receiver
+# ---------------------------------------------------------
+# WITH followees AS
+# (
+#     SELECT
+#         follows.followee_user_id AS followee_user_id
+#     FROM
+#         follows
+#     WHERE
+#         follows.is_current = true
+#         AND follows.is_delete = false
+#         AND follows.follower_user_id = 1
+# )
+# ,
+# tips AS
+# (
+#     SELECT
+#         user_tips.signature AS signature,
+#         user_tips.slot AS slot,
+#         user_tips.sender_user_id AS sender_user_id,
+#         user_tips.receiver_user_id AS receiver_user_id,
+#         user_tips.amount AS amount,
+#         user_tips.created_at AS created_at
+#     FROM
+#         user_tips
+#         LEFT OUTER JOIN
+#             followees AS followees_for_receiver
+#             ON user_tips.receiver_user_id = followees_for_receiver.followee_user_id
+#         LEFT OUTER JOIN
+#             followees AS followees_for_sender
+#             ON user_tips.sender_user_id = followees_for_sender.followee_user_id
+#     WHERE
+#         followees_for_receiver.followee_user_id IS NOT NULL
+#         OR followees_for_sender.followee_user_id IS NOT NULL
+#     ORDER BY
+#         user_tips.slot DESC LIMIT 100 OFFSET 0
```

Comment:
is it necessary to sort and limit again later on?

very nit - how did we decide slot vs created at? from my understanding, created_at is a bit more accurate but just curious.

Reply:
We definitely don't - actually this might break right due to the offset? Good catch.

Reply:
Also the reason we don't use `created_at` here is I'm timestamping this per DN, but good callout maybe I can use the date set on the transaction

Reply:
ooh yeah i think it could break because of the group by. 

i think generally we used created_at for transaction timestamp and updated_at for inserting into the DB.

Reply:
I should follow suit!! thanks, will make a follow up